content/movie-review-pt1.md

```diff
@@ -0,0 +1,450 @@
+        self.number
+    }
+}
+```
+You can call the implementations of the example struct like so:
+
+```rust
+let example = Example { 
+    number = 1 
+}
+
+example.boo();
+```
+
+### Traits and attributes
+
+You won't be creating your own traits or attributes at this stage, so we won't go into details about either. However, you will be using the `derive` attribute macro and some traits provided by the `borsh` crate, so it's important you have a high level understanding of each.
+
+Traits describe an abstract interface that types can implement. If a trait defines a function `bark()` and a type then adopts that trait, the type must then implement the `bark()` function.
+
+[Attributes](https://doc.rust-lang.org/rust-by-example/attribute.html) add metadata to a type and can be used for many different purposes.
+
+When you add the [`derive` attribute](https://doc.rust-lang.org/rust-by-example/trait/derive.html) to a type and provide one or more supported traits, code is generated under the hood to automatically implement the traits for that type. We'll provide a concrete example of this shortly.
+
+## Representing instructions as a Rust data type
+
+Now that we've covered the Rust basics, let's apply them to Solana programs.
+
+More often than not, programs will have more than one function. For example, you may have a program that acts as the backend for notes. Assume this program accepts instructions for creating a new note, updating an existing note, and deleting an existing note. 
+
+Since instructions have discrete types, they're usually a great fit for an enum data type.
+
+```rust
+enum NoteInstruction {
+    CreateNote {
+        title: String,
+        body: String,
+        id: u64
+    },
+    UpdateNote {
+        title: String,
+        body: String,
+        id: u64
+    },
+    DeleteNote {
+        id: u64
+    }
+}
+```
+
+Notice that each variant of the `NoteInstruction` enum comes with embedded data that will be used by the program to accomplish the taks of creating, updating, and deleting a note, respectively.
+
+## Deserialize instruction data
+
+Instruction data is passed to the program as a byte array, so you need a way to deterministically convert that array into an instance of the instruction enum type.
+
+In previous modules, we used Borsh for client-side serialization and deserialization. To use Borsh program-side, we use the `borsh` crate. This crate provides traits for `BorshDeserialize` and `BorshSerialize` that you can apply to your types using the `derive` attribute. 
+
+To make deserializing instruction data simple, you can create a struct representing the data and use the `derive` attribute to apply the `BorshDeserialize` trait to the struct. This implements the methods defined in `BorshDeserialize`, including the the `try_from_slice` method that we'll be using to deserialize the instruction data.
+
+Remember, the struct itself needs to match the structure of the data in the byte array.
```

Comment:
When we say "match the structure of the data" do we mean the set of `name: type` relationships, or does the order of declaration top to bottom matter, e.g. below we have top to bottom `id`, `title`, `body` and above we declare top to bottom, `title`,` body`, `id` - should they match?

Reply:
The enum order doesn't matter because we're setting those explicitly. The order in the struct *does* matter because the implementation of BorshDeserialize will basically look at the first field's type and go look at the byte array expecting the first thing it sees to be of that type, then do the same for the next one, etc.

So "match structure of the data in the byte array" means that the order you put in the struct is the order that you'll need people to serialize the fields into the byte array client-side before submitting the transaction